File: LNCollectionView/LNScrollView/Pulse/LNScrollViewPulseGenerator.m
//
//  LNScrollViewPulseGenerator.m
//  LNCollectionView
//
//  Created by Levison on 27.11.24.
//

#import "LNScrollViewPulseGenerator.h"

@interface LNScrollViewPulseGenerator ()

@property (nonatomic, assign) BOOL isOpen;

@end

@implementation LNScrollViewPulseGenerator

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.mass = 1.f;
        self.isOpen = NO;
    }
    return self;
}

- (void)setMass:(CGFloat)mass
{
    _mass = MAX(1.0, mass);
}

- (CGFloat)generate:(CGFloat)velocity
{
    if (!self.isOpen) {
        return velocity;
    }
    
    if (velocity <= 0) {
        return velocity;
    }
    if (self.delegate && [self.delegate respondsToSelector:@selector(generatorHasDetectedMomentum:)]) {
        LNScrollViewMomentum *momentum = [[LNScrollViewMomentum alloc] init];
        momentum.mass = self.mass;
        momentum.velocity = velocity;
        return [self.delegate generatorHasDetectedMomentum:momentum].velocity;
    }
    return velocity;
}

- (void)open
{
    self.isOpen = YES;
}

- (void)close
{
    self.isOpen = NO;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/scrollview/pulse/LNScrollViewPulseGenerator.kt
package com.ln.collectionview.scrollview.pulse

import com.ln.collectionview.scrollview.status.LNScrollViewAutoEffectDelegate
import java.lang.ref.WeakReference

interface LNScrollViewPulseGeneratorDelegate {
    fun generatorHasDetectedMomentum(momentum: LNScrollViewMomentum): LNScrollViewMomentum
}

class LNScrollViewPulseGenerator {

    private var isOpen: Boolean = false
    var mass: Float = 1f
        set(value) {
            field = if (value < 1.0f) 1.0f else value
        }

    var delegate: WeakReference<LNScrollViewPulseGeneratorDelegate>? = null

    init {
        mass = 1f
        isOpen = false
    }

    fun generate(velocity: Float): Float {
        if (!isOpen) {
            return velocity
        }

        if (velocity <= 0) {
            return velocity
        }

        delegate?.get()?.let {
            val momentum = LNScrollViewMomentum().apply {
                this.mass = this@LNScrollViewPulseGenerator.mass
                this.velocity = velocity
            }
            val resultMomentum = it.generatorHasDetectedMomentum(momentum)
            return resultMomentum.velocity
        }

        return velocity
    }

    fun openStatus(): Boolean {
        return isOpen
    }

    fun open() {
        isOpen = true
    }

    fun close() {
        isOpen = false
    }
}
